cpp/src/arrow/compute/exec/order_by_impl.cc

```diff
@@ -86,6 +86,45 @@ class SelectKBasicImpl : public SortBasicImpl {
   const SelectKOptions options_;
 };
 
+class FetchBasicImpl : public SortBasicImpl {
+ public:
+  FetchBasicImpl(ExecContext* ctx, const std::shared_ptr<Schema>& output_schema,
+                 int64_t offset, int64_t count, SortOptions sort_options)
+      : SortBasicImpl(ctx, output_schema),
+        offset_(offset),
+        count_(count),
+        sort_options_(sort_options) {}
+
+  Result<Datum> DoFinish() override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    ARROW_ASSIGN_OR_RAISE(auto table,
+                          Table::FromRecordBatches(output_schema_, std::move(batches_)));
+    if (sort_options_.sort_keys.size() > 0) {
+      ARROW_ASSIGN_OR_RAISE(auto indices,
+                            SortIndices(table, std::move(sort_options_), ctx_));
+      ARROW_ASSIGN_OR_RAISE(auto sorted_table,
+                            Take(table, indices, TakeOptions::NoBoundsCheck(), ctx_));
+      return sorted_table.table()->Slice(offset_, count_);
```

Comment:
It would be slightly more efficient to slice the sorted indices and then take the sliced sorted indices.

Reply:
Thanks for noting this. Changes included.